inductor: fix .to(memort_format) issue which doesn't generate right stride

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91948


Motivation: for **.to(memory_format),** the inductor doesn't generate the right stride, see the following example:
```
class Model(torch.nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        x = x.to(memory_format=torch.contiguous_format)
        return x
```

the generated code doesn't do the memory format change and gets a wrong stride **(802816, 1, 14336, 256)**, it is not a contiguous stride.

```
from ctypes import c_void_p, c_long
import torch
import random
from torch import empty_strided, as_strided, device
from torch._inductor.codecache import AsyncCompile

aten = torch.ops.aten
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
async_compile = AsyncCompile()


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    return (arg0_1, )


if __name__ == "__main__":
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((128, 256, 56, 56), (802816, 1, 14336, 256), device='cpu', dtype=torch.float32)
    print_performance(lambda: call([arg0_1]))
```

After this PR, the will have a memory format change:

```
from ctypes import c_void_p, c_long
import torch
import random
from torch import empty_strided, as_strided, device
from torch._inductor.codecache import AsyncCompile

aten = torch.ops.aten
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
async_compile = AsyncCompile()


kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_xiaobing/77/c7773nj5pwikpmm2pwa62rcudlf7p3if7eyqb5k4sjsvewwje4le.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       float* __restrict__ out_ptr0)
{
    #pragma omp parallel num_threads(40)
    {
        {
            #pragma omp for
            for(long i0=0; i0<128; i0+=1)
            {
                #pragma GCC ivdep
                for(long i1=0; i1<256; i1+=1)
                {
                    #pragma GCC ivdep
                    for(long i2=0; i2<3136; i2+=1)
                    {
                        auto tmp0 = in_ptr0[i1 + (256*i2) + (802816*i0)];
                        out_ptr0[i2 + (3136*i1) + (802816*i0)] = tmp0;
                    }
                }
            }
        }
    }
}
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    buf1 = empty_strided((128, 256, 56, 56), (802816, 3136, 56, 1), device='cpu', dtype=torch.float32)
    kernel_cpp_0(c_void_p(arg0_1.data_ptr()), c_void_p(buf1.data_ptr()))
    del arg0_1
    return (buf1, )


if __name__ == "__main__":
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((128, 256, 56, 56), (802816, 1, 14336, 256), device='cpu', dtype=torch.float32)
    print_performance(lambda: call([arg0_1]))
```




cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

XiaobingSuper committed:
inductor: fix .to(memort_format) issue which doesn't generate right stride

[ghstack-poisoned]

ngimel reviewed (changes requested):
cc @eellison, I thought #91185 fixed wrong strides at the graph boundary? Generally we don't want to force strides in the middle of inductor graph, see discussion in #90454. Besides, the decomposition that you are changing already produces correct strides, so you are forcing some unwanted behavior (namely, forcing strides on inductor) and you are introducing an inplace operator to decomposition where there should be no inplace op:
```
In [2]: a=torch.randn(2,3,4,5)

In [3]: b=torch._decomp.decompositions._to_copy(a, memory_format=torch.channels_last)

In [4]: b.stride()
Out[4]: (60, 1, 15, 3)

In [6]: c=torch._decomp.decompositions._to_copy(a, memory_format=torch.contiguous_format)

In [7]: c.stride()
Out[7]: (60, 20, 5, 1)
```

ngimel commented:
Hm yeah, `memory_format` shouldn't be channels_last

ngimel commented:
Ok, so when input doesn't require grad, IR is similar to what @eellison posted, copy is DCE'd and here we are:
```
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[2, 3, 4, 5]):
        # File: /scratch/ngimel/work/pytorch/../playground/repro_91948.py:10, code: x = x.to(memory_format=torch.contiguous_format)
        convert_element_type: f32[2, 3, 4, 5] = torch.ops.prims.convert_element_type.default(arg0_1, torch.float32);  arg0_1 = None
        empty_like: f32[2, 3, 4, 5] = torch.ops.aten.empty_like.default(convert_element_type, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False, memory_format = torch.contiguous_format)
        return (convert_element_type,)
```
this should be fixed by running functionalization always, what's stopping us? cc @Chillee 
But then when functionalization is run (by setting `x.requires_grad_(True)`) the copy is also eliminated:
```
class GraphModule(torch.nn.Module):
    def forward(self, primals_1: f32[2, 3, 4, 5]):
        # File: /scratch/ngimel/work/pytorch/../playground/repro_91948.py:10, code: x = x.to(memory_format=torch.contiguous_format)
        convert_element_type: f32[2, 3, 4, 5] = torch.ops.prims.convert_element_type.default(primals_1, torch.float32);  primals_1 = None
        return [convert_element_type]
```
repro script:
```
import torch
import torch._inductor.config as config
config.debug=True

class Model(torch.nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        x = x.to(memory_format=torch.contiguous_format)
        return x


mod = Model()
opt_mod = torch.compile(mod)
x=torch.randn(2,3,4,5).to(memory_format=torch.channels_last)#.requires_grad_(True)

out = opt_mod(x)
print(out.stride())
```
cc @bdhirsh

XiaobingSuper committed:
Update on "inductor: fix .to(memort_format) issue which doesn't generate right stride"



Motivation: for **.to(memory_format),** the inductor doesn't generate the right stride, see the following example:
```
class Model(torch.nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        x = x.to(memory_format=torch.contiguous_format)
        return x
```

the generated code doesn't do the memory format change and gets a wrong stride **(802816, 1, 14336, 256)**, it is not a contiguous stride.

```
from ctypes import c_void_p, c_long
import torch
import random
from torch import empty_strided, as_strided, device
from torch._inductor.codecache import AsyncCompile

aten = torch.ops.aten
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
async_compile = AsyncCompile()


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    return (arg0_1, )


if __name__ == "__main__":
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((128, 256, 56, 56), (802816, 1, 14336, 256), device='cpu', dtype=torch.float32)
    print_performance(lambda: call([arg0_1]))
```

After this PR, the will have a memory format change:

```
from ctypes import c_void_p, c_long
import torch
import random
from torch import empty_strided, as_strided, device
from torch._inductor.codecache import AsyncCompile

aten = torch.ops.aten
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
async_compile = AsyncCompile()


kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_xiaobing/77/c7773nj5pwikpmm2pwa62rcudlf7p3if7eyqb5k4sjsvewwje4le.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       float* __restrict__ out_ptr0)
{
    #pragma omp parallel num_threads(40)
    {
        {
            #pragma omp for
            for(long i0=0; i0<128; i0+=1)
            {
                #pragma GCC ivdep
                for(long i1=0; i1<256; i1+=1)
                {
                    #pragma GCC ivdep
                    for(long i2=0; i2<3136; i2+=1)
                    {
                        auto tmp0 = in_ptr0[i1 + (256*i2) + (802816*i0)];
                        out_ptr0[i2 + (3136*i1) + (802816*i0)] = tmp0;
                    }
                }
            }
        }
    }
}
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    buf1 = empty_strided((128, 256, 56, 56), (802816, 3136, 56, 1), device='cpu', dtype=torch.float32)
    kernel_cpp_0(c_void_p(arg0_1.data_ptr()), c_void_p(buf1.data_ptr()))
    del arg0_1
    return (buf1, )


if __name__ == "__main__":
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((128, 256, 56, 56), (802816, 1, 14336, 256), device='cpu', dtype=torch.float32)
    print_performance(lambda: call([arg0_1]))
```




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ngimel commented:
Changing decomp to return `torch.clone(x, memory_format=memory_format)`, which is close to what @XiaobingSuper has done, fixes it, so there are some shenanigans with aten.copy.default in the decomp, so I'm inclined to propose that as a fix (but we still need to run functionalization even if nothing requires grad).

XiaobingSuper commented:
@ngimel , thanks, currently, using **torch.clone(x, memory_format=memory_format)** to do the decomposition.

XiaobingSuper commented:
@pytorchbot merge
